libdevice/nativecpu_utils.cpp

```diff
@@ -37,15 +37,15 @@ DEVICE_EXTERN_C void __mux_work_group_barrier(uint32_t id, uint32_t scope,
                                               uint32_t semantics);
```

Comment:
```suggestion
                                              uint32_t semantics) noexcept;
```
And likewise below? It's okay to mark these as `noexcept`, they should not throw, and this avoids `__spirv_Control_Barrier` needing to set up an implicit `try { ... } catch { std::terminate(); }` around the call.

Reply:
I see there are many functions marked as DEVICE_EXTERN_C like __mux_work_group_barrier in this file. Can this be done separately since the change is big and isn't very closely related to the bug fix in this PR?

If I don't add except to below __spirv_ControlBarrier, there is build error:
libdevice/nativecpu_utils.cpp:39:1: error: '__spirv_ControlBarrier' is missing exception specification 'noexcept'

Reply:
I think the change is related to the bug fix in this PR: if the `noexcept` between `__mux_work_group_barrier` and `__spirv_Control_Barrier` matches, then `__spirv_Control_Barrier` compiles to a plain conditional `(tail) call` instruction. If the `noexcept` does not match, then `__spirv_Control_Barrier` needs to be compiled to more complicated code as I commented earlier. Currently, the `noexcept` does match so we get the simple code, and your change makes it so that it does not match and we would get the complicated code.

This does not need to be done for all the functions in this file, but the ones where you added the `noexcept` to the `__spirv_` version of the function, I think the `__mux_` versions of those functions do need to be updated too.

Reply:
thanks @hvdijk, you're right.

> This does not need to be done for all the functions in this file, but the ones where you added the noexcept to the __spirv_ version of the function, I think the __mux_ versions of those functions do need to be updated too.

Done, please double check again.

> We get this wrong in some other places already.

Yeah, I see places e.g. `GEN_u32(__spirv_SubgroupLocalInvocationId, __mux_get_sub_group_local_id)`, I didn't update them.